autogluon/utils/tabular/ml/models/tab_transformer/utils.py

```diff
@@ -0,0 +1,186 @@
+from tqdm import tqdm
+from torch.utils.data import Dataset, DataLoader
+from autogluon.utils.tabular.ml.models.tab_transformer.TabTransormerEncoder import WontEncodeError, NullEnc
+import torch
+from autogluon.utils.tabular.ml.models.tab_transformer import TabTransormerEncoder
+from autogluon.utils.tabular.ml.models.tab_transformer import pretexts
+
+
+def augmentation(data, target, mask_prob=0.4, num_augs=1):
+    shape=data.shape
+    cat_data=torch.cat([data for _ in range(num_augs)])
+    target=torch.cat([target for _ in range(num_augs)]).view(-1)
+    locs_to_mask = torch.empty_like(cat_data, dtype=float).uniform_() < mask_prob
+    cat_data[locs_to_mask] = 0 
+    cat_data=cat_data.view(-1,shape[-1])
+    return cat_data, target
+
+
+
+def epoch(net, data_loader, optimizers, loss_criterion, pretext, state, scheduler, epoch, epochs, aug_kwargs=None):    
+    is_train = (optimizers is not None)
+    net.train() if is_train else net.eval()
+    total_loss, total_correct, total_num, data_bar = 0.0, 0.0, 0, tqdm(data_loader)
+
+    if aug_kwargs is None:
+        aug_kwargs={'mask_prob': 0.4,
+                    'num_augs': 1}
+
+
+    with (torch.enable_grad() if is_train else torch.no_grad()):
+        for data, target in data_bar:
+            data, target = pretext.get(data, target)
+        
+            if state in [None, 'finetune']:
+                data, target = augmentation(data,target, **aug_kwargs)
+                out, _    = net(data)
+              
+            elif state=='pretrain':
+                _, out    = net(data)
+            else:
+                raise NotImplementedError("state must be one of [None, 'pretrain', 'finetune']")
+        
+            loss, correct  = pretext(out, target)
+   
+            if is_train:
+                for optimizer in optimizers:
+                    optimizer.zero_grad()
+                loss.backward()
+                for optimizer in optimizers:
+                    optimizer.step()
+    
+            total_num += 1
+            total_loss += loss.item() 
+
+            if epochs==1:
+                train_test = 'Test'
+            else:
+                train_test = 'Train'
+
+            if correct is not None:
+                total_correct += correct.mean().cpu().numpy() 
+                data_bar.set_description('{} Epoch: [{}/{}] Loss: {:.4f} Acc: {:.2f}%'.format(train_test, epoch, epochs, total_loss / total_num, total_correct / total_num * 100))
+            else:
+                data_bar.set_description('{} Epoch: [{}/{}] Loss: {:.4f}'.format(train_test, epoch, epochs, total_loss / total_num))
+        return total_loss / total_num, total_correct / total_num * 100
+
+    if scheduler is not None:
+        scheduler.step()
+    return total_loss / total_num
+
+
+def get_col_info(X):
+    cols=list(X.columns)
+    col_info=[]
+    for c in cols:
+        col_info.append({"name": c, "type": "CATEGORICAL"})
+    return col_info
+
+
+class TabTransformerDataset(Dataset):
```

Comment:
does this class provide duplicate functionality of `TabTransDataset`? Can we unify them?

Reply:
+1